src/operator/nn/cudnn/cudnn_batch_norm.cu

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * Copyright (c) 2015 by Contributors
+ * \file cudnn_batch_norm.cu
+ * \brief
+ * \author Junyuan Xie, Da Zheng
+*/
+
+#include "./cudnn_batch_norm-inl.h"
+#include <vector>
+
+namespace mxnet {
+namespace op {
+#if CUDNN_MAJOR == 4
+
+template<typename DType>
+static CuDNNBatchNormOp<DType> &GetCuDNNOp(const BatchNormParam& param) {
+  static thread_local CuDNNBatchNormOp<DType> op;
```

Comment:
Are we still going to use this thread_local method with static op rather than just call a static function? You aren't going to be able to save state in the class anyway because there's no guarantee that the same thread will pass back here for the forward and backward pass, or even all backwards passes for that matter. I feel this is misleading because people are going to try to save state, plus whatever overhead is involved with the thread_local (it's not zero for all platforms).

Reply:
For CuDNN operators, this is necessary because cudnn operators require a large number of states. Initializing these states has two phases: creating memory and setting values. Although we can't avoid setting values for these states in forward/backward, this design (with thread_local) can at least avoid creating memory over and over again. @piiswrong thinks this matters a lot. I'll add comments why we want to do this.

Reply:
So this makes the operator sensitive to which thread calls Forward and Backward?

Reply:
Also, so if CUDNN batch norm saves state in the class (for example), and if there is a network like:

(1)Convolution->(2)BatchNorm->(3)Activation->(4)Convolution->(5)BatchNorm->(6)Activation

...And one thread calls all of these in sequence, will this code mix up the state information between the first forward call of batch norm and the second?  Will the backward pass through (2) be using state information from the (5) since that was the last state saved (presumably from forward pass, but maybe also from backward pass)?

Reply:
no, it won't. The values of the state are only valid within a function (either forward or backward). The values of the state are always reinitialized at the beginning. The only thing we want to avoid is to create memory for the state again and again.